### PR TITLE
some header tweaks to speed up C++ IR tests a tad

### PR DESCRIPTION
**Pull request by catoverdrive**

### CODE REVIEW

**Comment by catoverdrive** [comment]
@cseed this is rebased and I think ready for a look now.

File: hail/src/main/resources/include/hail/Region.h [hunk 1]
```diff
@@ -3,141 +3,105 @@
...
-            return p;
-          } else {
-            return (n <= block_threshold) ? allocate_new_block(n) : allocate_big_chunk(n);
-          }
-        }
-        SharedPtr get_region();
-        void add_reference_to(SharedPtr region);
-    };
+    class SharedPtr {
+    friend class RegionPool;
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/Region.h]: Indent

File: hail/src/main/resources/include/hail/Region.h [hunk 2]
```diff
@@ -3,141 +3,105 @@
...
+    private:
+      Region * region_;
+      void clear();
+      explicit SharedPtr(Region * region) : region_(region) {
+      if (region_ != nullptr) { ++(region_->references_); }
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/Region.h]: Indent

File: hail/src/main/resources/include/hail/Region.h [hunk 3]
```diff
@@ -3,141 +3,105 @@
...
+    }
+    public:
+      SharedPtr(const SharedPtr &ptr) : SharedPtr(ptr.region_) { }
+      SharedPtr(SharedPtr &&ptr) : region_(ptr.region_) { ptr.region_ = nullptr; }
+      ~SharedPtr() { clear(); }
+
+      void swap(SharedPtr &other) { std::swap(region_, other.region_); }
+      SharedPtr& operator=(SharedPtr other) {
+      swap(other);
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/Region.h]: Indent

File: hail/src/main/resources/include/hail/Region.h [hunk 4]
```diff
@@ -3,141 +3,105 @@
...
+    int references_ {0};
+    size_t block_offset_;
+    std::unique_ptr<char[]> current_block_;
+    std::vector<std::unique_ptr<char[]>> used_blocks_{};
+    std::vector<std::unique_ptr<char[]>> big_chunks_{};
+    std::vector<SharedPtr> parents_{};
+    char * allocate_new_block(size_t n);
+    char * allocate_big_chunk(size_t size);
+    explicit Region(RegionPool * pool);
+    public:
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/Region.h]: Unindent

File: hail/src/main/resources/include/hail/Region.h [hunk 5]
```diff
@@ -3,141 +3,105 @@
...
+      Region(Region &pool) = delete;
+      Region(Region &&pool) = delete;
+      Region& operator=(Region pool) = delete;
+      void clear();
+      void align(size_t a) {
+      block_offset_ = (block_offset_ + a-1) & ~(a-1);
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/Region.h]: Indent

File: hail/src/main/resources/include/hail/Region.h [hunk 6]
```diff
@@ -3,141 +3,105 @@
...
+class ScalaRegion : public NativeObj {
+  public:
+    RegionPtr region_;
+    ScalaRegion(ScalaRegionPool * pool);
+    void align(size_t alignment) { region_->align(alignment); }
+    char * allocate(size_t alignment, size_t n) { return region_->allocate(alignment, n); }
+    char * allocate(size_t n) { return region_->allocate(n); }
+
+    virtual const char* get_class_name() { return "Region"; }
+    virtual ~ScalaRegion() { region_ = nullptr; }
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/Region.h]: You can use `virtual ~ScalaRegion() = default;`, which will destruct `region_`.

**Pull request merged**